File: src/main/java/org/jlab/ersap/actor/datatypes/CodaTimeFrameBinaryDataType.java
package org.jlab.ersap.actor.datatypes;

import org.jlab.epsci.ersap.base.error.ErsapException;
import org.jlab.epsci.ersap.engine.EngineDataType;
import org.jlab.epsci.ersap.engine.ErsapSerializer;
import org.jlab.ersap.actor.coda.proc.EtEvent;
import org.jlab.ersap.actor.coda.proc.FADCHit;
import org.jlab.ersap.actor.coda.proc.RocTimeFrameBank;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.*;

public class CodaTimeFrameBinaryDataType {

    public static final String MIME_TYPE = "binary/coda-time-frame";

    private static class CodaTimeFrameBinarySerializer implements ErsapSerializer {

        private static byte[] write(EtEvent event) throws IOException {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            DataOutputStream out = new DataOutputStream(baos);

            List<List<RocTimeFrameBank>> timeFrames = event.getTimeFrames();
            out.writeInt(timeFrames.size());

            for (List<RocTimeFrameBank> frame : timeFrames) {
                out.writeInt(frame.size());
                for (RocTimeFrameBank roc : frame) {
                    out.writeInt(roc.getRocID());
                    out.writeInt(roc.getFrameNumber());
                    out.writeLong(roc.getTimeStamp());

                    List<FADCHit> hits = roc.getHits();
                    out.writeInt(hits.size());

                    for (FADCHit h : hits) out.writeInt(h.crate());
                    for (FADCHit h : hits) out.writeInt(h.slot());
                    for (FADCHit h : hits) out.writeInt(h.channel());
                    for (FADCHit h : hits) out.writeInt(h.charge());
                    for (FADCHit h : hits) out.writeLong(h.time());
                }
            }

            return baos.toByteArray();
        }

        private static EtEvent read(byte[] data) throws IOException {
            ByteBuffer buf = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN);
            EtEvent event = new EtEvent();

            int tfCount = buf.getInt();
            for (int t = 0; t < tfCount; t++) {
                int rocCount = buf.getInt();
                List<RocTimeFrameBank> frame = new ArrayList<>();

                for (int r = 0; r < rocCount; r++) {
                    RocTimeFrameBank roc = new RocTimeFrameBank();
                    roc.setRocID(buf.getInt());
                    roc.setFrameNumber(buf.getInt());
                    roc.setTimeStamp(buf.getLong());

                    int hitCount = buf.getInt();
                    int[] crates = new int[hitCount];
                    int[] slots = new int[hitCount];
                    int[] chans = new int[hitCount];
                    int[] charges = new int[hitCount];
                    long[] times = new long[hitCount];

                    for (int i = 0; i < hitCount; i++) crates[i] = buf.getInt();
                    for (int i = 0; i < hitCount; i++) slots[i] = buf.getInt();
                    for (int i = 0; i < hitCount; i++) chans[i] = buf.getInt();
                    for (int i = 0; i < hitCount; i++) charges[i] = buf.getInt();
                    for (int i = 0; i < hitCount; i++) times[i] = buf.getLong();

                    for (int i = 0; i < hitCount; i++) {
                        roc.addHit(new FADCHit(crates[i], slots[i], chans[i], charges[i], times[i]));
                    }
                    frame.add(roc);
                }
                event.addTimeFrame(frame);
            }

            return event;
        }

        @Override
        public ByteBuffer write(Object data) throws ErsapException {
            if (!(data instanceof EtEvent)) {
                throw new ErsapException("Expected EtEvent, got " + data.getClass().getName());
            }
            try {
                byte[] binary = write((EtEvent) data);
                return ByteBuffer.wrap(binary);
            } catch (IOException e) {
                throw new ErsapException("Serialization failed", e);
            }
        }

        @Override
        public Object read(ByteBuffer buffer) throws ErsapException {
            try {
                byte[] binary = new byte[buffer.remaining()];
                buffer.get(binary);
                return read(binary);
            } catch (IOException e) {
                throw new ErsapException("Deserialization failed", e);
            }
        }
    }

    public static final EngineDataType CODA_TIME_FRAME =
            new EngineDataType(MIME_TYPE, new CodaTimeFrameBinarySerializer());
}

File: src/main/cpp/src/CodaTImeFrameBinaryDataType.cpp
// CodaTimeFrameBinaryDataType.cpp
#include "CodaTimeFrameBinaryDataType.hpp"
#include <stdexcept>
#include <cstring>
#include <cstdint>
#include <vector>

namespace ersap {
namespace coda {


std::vector<std::uint8_t> CodaTimeFrameSerializer::serializeToBinary(const CodaTimeFrame& event) {
    std::vector<std::uint8_t> buffer;

    auto writeInt32 = [&](std::int32_t val) {
        for (int i = 0; i < 4; ++i)
            buffer.push_back(static_cast<uint8_t>((val >> (i * 8)) & 0xFF));
    };

    auto writeInt64 = [&](std::int64_t val) {
        for (int i = 0; i < 8; ++i)
            buffer.push_back(static_cast<uint8_t>((val >> (i * 8)) & 0xFF));
    };

    const auto& timeFrames = event.timeFrames;
    writeInt32(static_cast<std::int32_t>(timeFrames.size()));

    for (const auto& tf : timeFrames) {
        writeInt32(static_cast<std::int32_t>(tf.size()));

        for (const auto& roc : tf) {
            writeInt32(roc.rocId);
            writeInt32(roc.frameNumber);
            writeInt64(roc.timeStamp);

            const auto& hits = roc.hits;
            writeInt32(static_cast<std::int32_t>(hits.size()));

            for (const auto& h : hits) writeInt32(h.crate);
            for (const auto& h : hits) writeInt32(h.slot);
            for (const auto& h : hits) writeInt32(h.channel);
            for (const auto& h : hits) writeInt32(h.charge);
            for (const auto& h : hits) writeInt64(h.time);
        }
    }

    return buffer;
}

CodaTimeFrame CodaTimeFrameSerializer::deserializeFromBinary(const std::vector<std::uint8_t>& buffer) {
    size_t offset = 0;

    auto readInt32 = [&]() -> std::int32_t {
        if (offset + 4 > buffer.size()) throw std::runtime_error("Underflow int32");
        std::int32_t val = 0;
        for (int i = 0; i < 4; ++i)
            val |= (buffer[offset++] << (i * 8));
        return val;
    };

    auto readInt64 = [&]() -> std::int64_t {
        if (offset + 8 > buffer.size()) throw std::runtime_error("Underflow int64");
        std::int64_t val = 0;
        for (int i = 0; i < 8; ++i)
            val |= (static_cast<std::int64_t>(buffer[offset++]) << (i * 8));
        return val;
    };

    CodaTimeFrame event;
    int tfCount = readInt32();

    for (int t = 0; t < tfCount; ++t) {
        TimeFrame tf;
        int rocCount = readInt32();

        for (int r = 0; r < rocCount; ++r) {
            RocTimeFrameBank roc;
            roc.rocId = readInt32();
            roc.frameNumber = readInt32();
            roc.timeStamp = readInt64();

            int hitCount = readInt32();
            std::vector<std::int32_t> crates(hitCount), slots(hitCount), chans(hitCount), charges(hitCount);
            std::vector<std::int64_t> times(hitCount);

            for (int i = 0; i < hitCount; ++i) crates[i] = readInt32();
            for (int i = 0; i < hitCount; ++i) slots[i] = readInt32();
            for (int i = 0; i < hitCount; ++i) chans[i] = readInt32();
            for (int i = 0; i < hitCount; ++i) charges[i] = readInt32();
            for (int i = 0; i < hitCount; ++i) times[i] = readInt64();

            for (int i = 0; i < hitCount; ++i)
                roc.addHit(FADCHit{crates[i], slots[i], chans[i], charges[i], times[i]});

            tf.push_back(std::move(roc));
        }
        event.addTimeFrame(tf);
    }

    return event;
}

// CodaTimeFrameSerializer implementation
std::vector<std::uint8_t> CodaTimeFrameSerializer::write(const ersap::any& data) const {
    const auto& event = ersap::any_cast<const CodaTimeFrame&>(data);
    return CodaTimeFrameSerializer::serializeToBinary(event);
}

ersap::any CodaTimeFrameSerializer::read(const std::vector<std::uint8_t>& buffer) const {
    CodaTimeFrame event = CodaTimeFrameSerializer::deserializeFromBinary(buffer);
    return ersap::any{std::move(event)};
}

} // namespace coda
} // namespace ersap
